src/languages/fsharp.js

```diff
@@ -84,63 +78,368 @@ export default function(hljs) {
+          /\\(?:.|\d{3}|x[a-fA-F\d]{2}|u[a-fA-F\d]{4}|U[a-fA-F\d]{8})/ // ...or an escape sequence
+        ),
+      /'/
+    ),
+    relevance: 0
+  };
+  SUBST.contains = [
+    INTERPOLATED_VERBATIM_STRING,
+    INTERPOLATED_STRING,
+    VERBATIM_STRING,
+    FSHARP_CHAR_LITERAL,
+    QUOTED_STRING,
+    hljs.C_NUMBER_MODE,
+    FSHARP_COMMENT
+  ];
+  const FSHARP_STRING = {
+    variants: [
+      INTERPOLATED_TRIPLE_QUOTED_STRING,
+      INTERPOLATED_VERBATIM_STRING,
+      INTERPOLATED_STRING,
+      VERBATIM_STRING,
+      TRIPLE_QUOTED_STRING,
+      QUOTED_STRING,
+      FSHARP_CHAR_LITERAL
+    ]
+  };
+
+  // 'a or ^a
+  const GENERIC_TYPE_SYMBOL = {
+    match: regex.concat(/('|\^)/, hljs.UNDERSCORE_IDENT_RE),
+    scope: 'symbol',
+    relevance: 0
+  };
+
   return {
     name: 'F#',
     aliases: [
       'fs',
       'f#'
     ],
-    keywords: KEYWORDS,
+    keywords: {
+      type: TYPES,
+      keyword: KEYWORDS,
+      literal: LITERALS,
+      built_in: BUILTINS
+    },
     illegal: /\/\*/,
     contains: [
       {
-        // monad builder keywords (matches before non-bang kws)
-        className: 'keyword',
-        begin: /\b(yield|return|let|do)!/
+        // monad builder keywords (matches before non-bang keywords)
+        scope: 'keyword',
+        match: /\b(yield|return|let|do|match|use)!/
       },
+      FSHARP_STRING,
+      FSHARP_COMMENT,
       {
-        className: 'string',
-        begin: '@"',
-        end: '"',
+        // type MyType<'a> = ...
+        begin: [
+          /type/,
+          /\s+/,
+          hljs.UNDERSCORE_IDENT_RE
+        ],
+        beginScope: {
+          1: 'keyword',
+          3: 'title.class'
+        },
+        end: /\(|=|$/,
+        excludeEnd: true,
         contains: [
-          {
-            begin: '""'
-          }
+          GENERIC_TYPE_SYMBOL
         ]
       },
       {
-        className: 'string',
-        begin: '"""',
-        end: '"""'
+        // computation expressions:
+        begin: [
+            /\b[_a-z]\w*/,
+            /(?=\s*\{)/
+          ],
+        beginScope: { 1: 'keyword' }
       },
-      hljs.COMMENT('\\(\\*(\\s)', '\\*\\)', {
-        contains: ["self"]
-      }),
       {
-        className: 'class',
-        beginKeywords: 'type',
-        end: '\\(|=|$',
+        // preprocessor directives and fsi commands:
+        begin: [
+          /^\s*/,
+          regex.concat(/#/, regex.either(...PREPROCESSOR_KEYWORDS)),
+          /\b/
+        ],
+        beginScope: { 2: 'meta' },
+        end: /\s|$/,
         excludeEnd: true,
-        contains: [
-          hljs.UNDERSCORE_TITLE_MODE,
-          TYPEPARAM
-        ]
+        relevance: 0
       },
       {
-        className: 'meta',
-        begin: '\\[<',
-        end: '>\\]',
-        relevance: 10
+        // [<Attributes("")>]
+        scope: 'meta',
+        begin: /^\s*\[<(?=[<\w])/,
+        excludeBegin: true,
+        end: />\]/,
+        excludeEnd: true,
```

Comment:
I can also replace this `excludeEnd` with a look-ahead in the `end` regex, what do you think?

Should I also replace the `excludeBegin` with a `begin` regex in a look behind like so (with a look ahead for the expected content): `/(?<=^\s*\[<)(?=\s*\w)/`?

(and I just noticed the current regex has an extra `<` in the content look ahead that I inadvertently added when copy pasting from the csharp version...)

Reply:
I went ahead and resolved these in my commit.

Reply:
Yes, I saw, but what about the `excludeBegin`? Should we also strive to replace them with look behinds?

Reply:
> Should we also strive to replace them with look behinds?

No, because _we can't use look-behinds at all yet_ - see prior discussion.   When we support look-behind that's definitely something we'd have a look at.

Reply:
> we can't use look-behinds at all yet

Ah I did not understand that, I thought it was only *negative* look behind we couldn't use, my bad.

Why does [the doc](https://highlightjs.readthedocs.io/en/latest/language-guide.html#regular-expression-features) seems to indicate look-behinds in `begin` are supported?

Reply:
The key is in the wording:

> look-behind matching (**when JS supports it**) for begin (#2135)

The key being "when JS [browsers] supports it".  _Nothing in our engine prevents them from working for begin - if the JS engine includes support._ (as you've seen when using them initially)   If you wanted to fork the library or use a 3rd party grammar that uses them - go for it - you just lose support for Mac OS Safari.  At least one 3rd party library does feature detection - resulting in inconsistent behavior across browsers - but thats' not a tradeoff we're interested in for Core.

However,  the engine itself would break if look-behind was used in `end` because of the way our engine is currently designed - the end matchers do not always have access to the entire string.

Reply:
Ok. I'm still a bit confused by the subtlety, but ok ^^

Reply:
https://github.com/highlightjs/highlight.js/blob/dea3cde209d0fe5a164607a61da8c42d796fa459/docs/language-guide.rst#regular-expression-features